p2pserver/message/utils/msg_workerpool.go

```diff
@@ -0,0 +1,232 @@
+	this.maxWorkerIdleTime = 10 * time.Second
+}
+
+var msgJobChanCap = func() int {
+	// Use blocking msgJobChan if GOMAXPROCS=1.
+	// This immediately switches receiveMsg to disposeMsg, which results
+	// in higher performance
+	if runtime.GOMAXPROCS(0) == 1 {
+		return 0
+	}
+
+	// Use non-blocking msgJobChan if GOMAXPROCS>1,
+	// since otherwise the receiveMsg caller (Acceptor) may lag accepting
+	// new message if disposeMsg is CPU-bound.
+	return 1
+}()
+
+func (this* msgWorkerPool) start() {
+
+	if this.waitingWokers == nil || this.stopChan == nil{
+		log.Error("[p2p]invalid start invoking, the msg worker pool hasn't been initialized")
+		return
+	}
+
+	go func() {
+		for {
+			this.clean()
+			select {
+			case <- this.stopChan:
+				return
+			default:
+				time.Sleep(this.maxWorkerIdleTime)
+			}
+		}
+	}()
+}
+
+func (this* msgWorkerPool) stop() {
+
+	if this.stopChan == nil {
+		log.Error("[p2p]invalid stop invoking, the msg worker pool hasn't been initialized!")
+		return
+	}
+	close(this.stopChan)
+	this.stopChan = nil
+}
+
+func (this* msgWorkerPool) clean() {
+
+	var willCleanJC []*msgJobChan
+	curTime := time.Now()
+
+	this.lock.Lock()
+	for _, v := range this.waitingWokers {
+		jobChs := *v
+		n := len(jobChs)
+		if n <= 0 {
+			continue
+		}
+		i := 0
+		for i < n  && jobChs[i] != nil && curTime.Sub(jobChs[i].lastUseTime) > this.maxWorkerIdleTime {
+			i++
+		}
+		willCleanJC = append(willCleanJC[:0], jobChs[:i]...)
+		if i > 0 {
+			m := copy(*v, jobChs[i:])
+			for j := m; j < n; j++ {
+				(*v)[j] = nil
+			}
+			*v = (*v)[:m]
+		}
+	}
+	this.lock.Unlock()
+
+	tmp := willCleanJC
+	for i, jCh := range tmp {
+		jCh.jobChan <- nil
+		tmp[i] = nil
+	}
+}
+
+func (this* msgWorkerPool) getMsgWorkChan(msgType string)*msgJobChan{
+
+	var msgJobCh *msgJobChan = nil
+	var willCreateNew = false
+
+	this.lock.Lock()
+	defer this.lock.Unlock()
+	if msgWaitingWorks, ok := this.waitingWokers[msgType]; ok {
+		lmWW := len(*msgWaitingWorks)
+		if lmWW > 0 {
+			msgJobCh = (*msgWaitingWorks)[lmWW-1]
+			(*msgWaitingWorks)[lmWW-1] = nil
+			*msgWaitingWorks = (*msgWaitingWorks)[:(lmWW-1)]
+		}else {
+			if this.curWorkerCount < this.maxWorkerCount {
+				willCreateNew = true
+			}
+		}
+	}else {
+		msgWaitingWorksT            := make([]*msgJobChan, 0)
+		this.waitingWokers[msgType] = &msgWaitingWorksT
+
+		if this.curWorkerCount < this.maxWorkerCount {
+			willCreateNew = true
+		}
+	}
+
+	if msgJobCh == nil && willCreateNew{
+		mJobCh := this.workChanPool.Get()
+
+		if mJobCh == nil {
+			mJobCh = &msgJobChan{
+				jobChan: make(chan *msgJobItem, msgJobChanCap),
+			}
+		}
+		msgJobCh = mJobCh.(*msgJobChan)
+
+		this.curWorkerCount++
+		go func() {
+			this.disposeMsg(msgJobCh)
+			this.workChanPool.Put(msgJobCh)
+		}()
+	}
+
+	return msgJobCh
+}
+
+func (this* msgWorkerPool) receiveMsg(msg *msgJobItem) bool {
+
+	msgJobCh:= this.getMsgWorkChan(msg.msgPayload.Payload.CmdType())
+	if msgJobCh != nil {
+		msgJobCh.jobChan <- msg
```

Comment:
should be lock-free in this function.
drop msg if jobChan is full.

Reply:
You are right. I have modify it.

Reply:
I don't think you get the point. `receiveMsg` can not be blocked since it will cause `MessageRouter` stall. what if `jobChan` is full?

Reply:
The msg woker pool's purpose is to limit vast msg specific disposing.  If the block sync msg is not disposed , the MessageRouter only discard it. In my opion, this won't cause MessageRouter stall.
     In receiveMsg, once all jobChan are using,  getMsgWorkChan will return nil and it won't affect MessageRouter  disposing other msgs.